pkg/controlplane/controller/defaultservice/default_service_controller.go

```diff
@@ -0,0 +1,337 @@
+func (c *Controller) sync() error {
+	if err := c.createOrUpdateMasterServiceIfNeeded(); err != nil {
+		return err
+	}
+	endpointPorts := createEndpointPortSpec(c.endpointPort, "https")
+	if err := c.endpointReconciler.ReconcileEndpoints(serviceName, c.endpointIP, endpointPorts, true); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createPortAndServiceSpec creates an array of service ports.
+// If the NodePort value is 0, just the servicePort is used, otherwise, a node port is exposed.
+func createPortAndServiceSpec(servicePort int, targetServicePort int, nodePort int, servicePortName string) ([]v1.ServicePort, v1.ServiceType) {
+	// Use the Cluster IP type for the service port if NodePort isn't provided.
+	// Otherwise, we will be binding the API server service to a NodePort.
+	servicePorts := []v1.ServicePort{{
+		Protocol:   v1.ProtocolTCP,
+		Port:       int32(servicePort),
+		Name:       servicePortName,
+		TargetPort: intstr.FromInt(targetServicePort),
+	}}
+	serviceType := v1.ServiceTypeClusterIP
+	if nodePort > 0 {
+		servicePorts[0].NodePort = int32(nodePort)
+		serviceType = v1.ServiceTypeNodePort
+	}
+	return servicePorts, serviceType
+}
+
+// createEndpointPortSpec creates the endpoint ports
+func createEndpointPortSpec(endpointPort int, endpointPortName string) []v1.EndpointPort {
+	return []v1.EndpointPort{{
+		Protocol: v1.ProtocolTCP,
+		Port:     int32(endpointPort),
+		Name:     endpointPortName,
+	}}
+}
+
+// createOrUpdateMasterServiceIfNeeded will create the specified service if it doesn't already exist.
+// TODO: rename to avoid use of term "master"
+func (c *Controller) createOrUpdateMasterServiceIfNeeded() error {
+	servicePorts, serviceType := createPortAndServiceSpec(c.servicePort, c.endpointPort, c.serviceNodePort, "https")
+	singleStack := v1.IPFamilyPolicySingleStack
+	svc := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceName,
+			Namespace: metav1.NamespaceDefault,
+			Labels:    map[string]string{"provider": "kubernetes", "component": "apiserver"},
+		},
+		Spec: v1.ServiceSpec{
+			Ports: servicePorts,
+			// maintained by this code, not by the pod selector
+			Selector:        nil,
+			ClusterIP:       c.serviceIP.String(),
+			IPFamilyPolicy:  &singleStack,
+			SessionAffinity: v1.ServiceAffinityNone,
+			Type:            serviceType,
+		},
+	}
+
+	s, err := c.serviceLister.Services(metav1.NamespaceDefault).Get(serviceName)
+	// Service does not exist
+	if err != nil && errors.IsNotFound(err) {
+		_, err := c.client.CoreV1().Services(metav1.NamespaceDefault).Create(context.TODO(), svc, metav1.CreateOptions{})
+		return err
+	}
+	// unexpected error
+	if err != nil {
+		return err
+	}
+	// Service exists; warn if reconciliation is necessary
+	// The Service kubernetes.default is never reconciled once created since
+	// https://github.com/kubernetes/kubernetes/commit/2a9a9fa155e92d0bd7f4eb85d843cfa14b830cca
+	// TODO: revisit if is necessary to change this behavior, add
```

Comment:
hmm - before your change we were reconciling service only on start

But with your change we're never reconciling it - seems like a bug...

Reply:
I'm rolling this changes back